File: MAP/LAB7-MAP/LAB7-MAP/Expressions/ArithmeticExpr.cs
using MAP_Lab7.Exceptions;
using MAP_Lab7.Model;
using System;

namespace MAP_Lab7.Expressions
{
    public class ArithmeticExpr : IExpression
    {
        private IExpression left;
        private IExpression right;
        private char op;

        public ArithmeticExpr(char o, IExpression l, IExpression r)
        {
            left = l;
            right = r;
            op = o;
        }

        public int Eval(IDictionaryy<string, int> dict)
        {
            try
            {
                int l = left.Eval(dict);
                int r = right.Eval(dict);

                if (op == '+')
                    return l + r;
                if (op == '-')
                    return l - r;
                if (op == '*')
                    return l * r;
                else
                {
                    if (r == 0)
                        throw new DivByZeroException("Division by zero !");
                    else
                        return l / r;
                }
                throw new InexistentSymbolException("Inexistent symbol error !");
            }
            catch (DivByZeroException e)
            {
                Console.WriteLine(e.Message);
            }
            catch (EvaluationException e2)
            {
                Console.WriteLine(e2.Message);
            }
            catch (InexistentSymbolException e3)
            {
                Console.WriteLine(e3.Message);
            }
            return -1;
        }

        public override string ToString()
        {
            return "" + left + ' ' + op + ' ' + right;
        }
    }
}

File: MAP/Lab3/src/Model/Expression/ArithmeticExpr.java
package Model.Expression;

import Exception.DivByZeroException;
import Exception.EvaluationException;
import Exception.InexistentSymbolException;
import Model.IDictionary;

public class ArithmeticExpr implements Expression
{
    private char operator;
    private Expression left;
    private  Expression right;

    public ArithmeticExpr(char op, Expression l, Expression r)
    {
        operator = op;
        left = l;
        right = r;
    }

    @Override
    public int Eval(IDictionary<String, Integer> d) throws DivByZeroException, InexistentSymbolException
    {
        try
        {
            int l = left.Eval(d);
            int r = right.Eval(d);
            if (operator == '+')
                return l + r;
            else if(operator == '-')
                return l - r;
            else if(operator == '*')
                return l * r;
            else if(operator == '/')
                if(r == 0)
                    throw new DivByZeroException("\nDivision by zero in " + l + operator + r + " ! \n");
                else return l / r;
            throw new InexistentSymbolException("\nSymbol " + operator + " does not exist !\n");
        }
        catch(EvaluationException | DivByZeroException | InexistentSymbolException e)
        {
            System.out.println(e.getMessage());
            return -1;
        }
    }

    @Override
    public String toString()
    {
        return left + " " + operator + " " + right ;
    }
}
